lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/stage/EmitDocsToDeleteByPrefix.java

```diff
@@ -0,0 +1,149 @@
+ * Config Parameters:
+ * - dropOriginal (Boolean, Optional): if set to true, will drop documents in the pipeline that were marked for deletion
+ *   as dropped. Does not apply to emitted documents, only the original documents. Defaults to false
+ * - addPrefix (String, optional) : String that adds on to all document ids used for listing
+ *    e.g. if id of doc is "doc1" & addPrefix is "-v1", then this stage will retrieve and emit documents where their
+ *         id prefix is "doc1-v1".
+ * - index (String) : name of the index to search within
+ * - apiKey (String) : apiKey to send requests
+ * - namespaces (Map<String, Object>, Optional) : mapping of namespaces of which you would like to collect ids on. If not given, will
+ *    retrieve from default namespace.
+ *    note that value of mapping will not matter, rather makes it convenient for user to reuse same configs as PineconeIndexer
+ * - deletionMarkerField (String) : name of the field which indicates the document is marked for deletion
+ * - deletionMarkerFieldValue (String) : value of the field which indicates that the document is marked for deletion
+ *   - for stage to work, deletionMarkerField and deletionMarkerFieldValue must be the same value provided indexer configs
+ */
+public class EmitDocsToDeleteByPrefix extends Stage {
+
+  private Pinecone client;
+  private final String indexName;
+  private Index index;
+  private final Map<String, Object> namespaces;
+  private final boolean dropOriginal;
+  private final String addPrefix;
+  private final String deletionMarkerField;
+  private final String deletionMarkerFieldValue;
+
+  public EmitDocsToDeleteByPrefix(Config config) {
+    super(config, new StageSpec()
+        .withOptionalProperties("dropOriginal", "addPrefix")
+        .withOptionalParents("namespaces")
+        .withRequiredProperties("apiKey", "deletionMarkerField", "deletionMarkerFieldValue", "index"));
+    this.indexName = config.getString("index");
+    this.namespaces = config.hasPath("namespaces") ? config.getConfig("namespaces").root().unwrapped() : null;
+    this.dropOriginal = config.hasPath("dropOriginal") ? config.getBoolean("dropOriginal") : false;
+    this.addPrefix = config.hasPath("addPrefix") ? config.getString("addPrefix") : "";
+    this.deletionMarkerField = config.getString("deletionMarkerField");
+    this.deletionMarkerFieldValue = config.getString("deletionMarkerFieldValue");
+  }
+
+  @Override
+  public void start() throws StageException {
+    try {
+      client = PineconeManager.getClient(config.getString("apiKey"));
+      index = client.getIndexConnection(indexName);
+    } catch (Exception e) {
+      throw new StageException("Unable to get client or index connection", e);
+    }
+
+    StateEnum state = client.describeIndex(indexName).getStatus().getState();
+    boolean isStable = state != StateEnum.INITIALIZING &&
+        state != StateEnum.INITIALIZATIONFAILED &&
+        state != StateEnum.TERMINATING;
+    if (!isStable) {
```

Comment:
Since this is the same check as is performed in PineconeIndexer, can we abstract these out to a method on the PineconeManager boolean isStable() (or similar) which returns the status of this. That way if Pinecone adds some new state down the line we only have to fix in one place.